src/renderers/dom/shared/__tests__/ReactServerRendering-test.js

```diff
@@ -23,20 +23,132 @@ var ReactTestUtils;
+  return expectErrors(
+    () => new Promise((resolve) => ReactDOM.render(reactElement, domElement, () => resolve(domElement.firstChild))),
+    errorCount
+  );
+}
+
+// Renders text using SSR and then stuffs it into a DOM node; returns the DOM
+// element that corresponds with the reactElement.
+// Does not render on client or perform client-side revival.
+function serverRender(reactElement, errorCount = 0) {
+  return expectErrors(
+    () => Promise.resolve(ReactDOMServer.renderToString(reactElement)),
+    errorCount)
+  .then((markup) => {
+    var domElement = document.createElement('div');
+    domElement.innerHTML = markup;
+    return domElement.firstChild;
+  });
+}
+
+const clientCleanRender = (element, errorCount = 0) => {
+  const div = document.createElement('div');
+  return renderIntoDom(element, div, errorCount);
+};
+
+const clientRenderOnServerString = (element, errorCount = 0) => {
+  return serverRender(element, errorCount).then((markup) => {
+    resetModules();
+    var domElement = document.createElement('div');
+    domElement.innerHTML = markup;
+    return renderIntoDom(element, domElement, errorCount);
+  });
+};
+
+const clientRenderOnBadMarkup = (element, errorCount = 0) => {
+  var domElement = document.createElement('div');
+  domElement.innerHTML = '<div id="badIdWhichWillCauseMismatch" data-reactroot="" data-reactid="1"></div>';
+  return renderIntoDom(element, domElement, errorCount + 1);
+};
+
+// runs a DOM rendering test as four different tests, with four different rendering
+// scenarios:
+// -- render to string on server
+// -- render on client without any server markup "clean client render"
+// -- render on client on top of good server-generated string markup
+// -- render on client on top of bad server-generated markup
+//
+// testFn is a test that has one arg, which is a render function. the render
+// function takes in a ReactElement and an optional expected error count and
+// returns a promise of a DOM Element.
+//
+// You should only perform tests that examine the DOM of the results of
+// render; you should not depend on the interactivity of the returned DOM element,
+// as that will not work in the server string scenario.
+function itRenders(desc, testFn) {
+  it(`${desc} with server string render`,
+    () => testFn(serverRender));
+  itClientRenders(desc, testFn);
+}
+
+// run testFn in three different rendering scenarios:
+// -- render on client without any server markup "clean client render"
+// -- render on client on top of good server-generated string markup
+// -- render on client on top of bad server-generated markup
+//
+// testFn is a test that has one arg, which is a render function. the render
+// function takes in a ReactElement and an optional expected error count and
+// returns a promise of a DOM Element.
+//
+// Since all of the renders in this function are on the client, you can test interactivity,
+// unlike with itRenders.
+function itClientRenders(desc, testFn) {
```

Comment:
None of the tests in this PR use it, but some of the ones coming later do. 

The idea is that `itRenders` is for tests that only need to inspect the DOM, whereas `itClientRenders` is for tests that need to test interactivity (things like events, refs, etc.). The latter tests are only possible when `ReactDOM.render` has been called, which doesn't happen in the server render-only tests.

I tried to explain that in the comment above the function, but obviously it wasn't clear enough. Any suggestions on how to make it clearer?